lib/beacon/broadcast.go

```diff
@@ -7,84 +7,26 @@
 package beacon
 
 import (
-	"fmt"
 	"net"
 	"time"
 
-	"github.com/thejerf/suture"
-
 	"github.com/syncthing/syncthing/lib/util"
 )
 
-type Broadcast struct {
-	*suture.Supervisor
-	port   int
-	inbox  chan []byte
-	outbox chan recv
-	br     *broadcastReader
-	bw     *broadcastWriter
-}
-
-func NewBroadcast(port int) *Broadcast {
-	b := &Broadcast{
-		Supervisor: suture.New("broadcastBeacon", suture.Spec{
-			// Don't retry too frenetically: an error to open a socket or
-			// whatever is usually something that is either permanent or takes
-			// a while to get solved...
-			FailureThreshold: 2,
-			FailureBackoff:   60 * time.Second,
-			// Only log restarts in debug mode.
-			Log: func(line string) {
-				l.Debugln(line)
-			},
-			PassThroughPanics: true,
-		}),
-		port:   port,
-		inbox:  make(chan []byte),
-		outbox: make(chan recv, 16),
-	}
-
-	b.br = &broadcastReader{
-		port:   port,
-		outbox: b.outbox,
-	}
-	b.br.ServiceWithError = util.AsServiceWithError(b.br.serve)
-	b.Add(b.br)
-	b.bw = &broadcastWriter{
-		port:  port,
-		inbox: b.inbox,
-	}
-	b.bw.ServiceWithError = util.AsServiceWithError(b.bw.serve)
-	b.Add(b.bw)
-
-	return b
-}
-
-func (b *Broadcast) Send(data []byte) {
-	b.inbox <- data
+func NewBroadcast(port int) Interface {
+	c := newCast("broadcastBeacon")
+	c.addReader(util.AsServiceWithError(func(stop chan struct{}) error {
+		return writeBroadcasts(c.inbox, port, stop)
+	}))
+	c.addWriter(util.AsServiceWithError(func(stop chan struct{}) error {
+		return readBroadcasts(c.outbox, port, stop)
+	}))
+	return c
```

Comment:
This could just take a function, and the addreader/writer wraps it in a service and passes in the right channel, and potentially a port (newCast could take a port)

Reply:
I did not unify this part as the function signatures vary a lot (different channels and port integer vs addr string) - passing around `util.ServiceWithErrors` avoids those problems.